extensions/multiline.md

```diff
@@ -0,0 +1,245 @@
+
+* `MULTILINE_MAX_BYTES <limit>`: `max-bytes` limit exceeded
+* `MULTILINE_MAX_LINES <limit>`: `max-lines` limit exceeded
+* `MULTILINE_INVALID_TARGET <batch-target> <provided-target>`: mismatched PRIVMSG target within a batch
+* `MULTILINE_INVALID`: any other error
+
+## Interaction with other specifications
+
+### Message tags ([spec][message tags])
+
+Clients MUST NOT send tags other than `draft/multiline-concat` and `batch` on messages within the batch. In particular, all client-only tags associated with the message must be sent attached to the initial BATCH command.
+
+### Message ids ([spec][message ids])
+
+Servers MUST only include a message ID on the first message of a batch when sending a fallback to non-supporting clients.
+
+### Labeled response ([draft][labeled response]) with echo-message ([spec][echo message])
+
+Servers MUST only include the label tag on the opening BATCH command when replying to a client using `echo-message`.
+
+## Client implementation considerations
+
+This section is non-normative.
+
+### Splitting long lines
+
+When using `draft/multiline-concat` to split longer lines, care is required to ensure messages are displayed appropriately when combined or in the fallback case.
+
+The recommended splitting method is to split long lines between words where possible and leave the space character at the end of the line. This ensures that pairs of words spanning a line concatenation are still space separated, without adding a space to the start of any fallback lines.
+
+In order to calculate the maximum allowed space before splitting a line, consider the following line:
+
+```
+:nick!~user@host PRIVMSG #channel :hello \r\n
+```
+
+The parts of a message that must fit within the 512 byte limit are as follows
+
+| part | syntax | length |
+|-|-|-|
+| leading | `:` | 1 |
+| nick | `nick` | variable |
+| nick/mask separator | `!` | 1 |
+| user | `~user` | variable |
+| user/host separator | `@` | 1 |
+| host | `host` | variable |
+| command | `PRIVMSG` | 7 |
+| channel | `#channel` | variable |
+| message separator | ` :` | 2 |
+| crlf | `\r\n` | 2 |
+| message | `hello ` | variable |
+
+Not all parts may appear in a message, but a safe limit should take them into account and include an extra safety buffer in case of any unexpected syntax added by a server, e.g. a statusmsg prefix.
+
+With a safety buffer of 10, the total message size available before a split is needed can be calculated as follows
+
+```
+512 - 14 - 10 - <nick> - <user> - <host> - <channel>
+```
+
+Clients might wish to perform a `WHO` query on themselves when connecting to a server, and keep track of numeric `396 RPL_VISIBLEHOST` to keep their `user@host` mask in sync.
+
+Clients might alternatively prefer to use a hard coded limit. A possible worst case scenario for the variable parts might be:
+
+* nick: 20
+* user: 20
+* host: 63
+* channel: 32
+
+Which would leave 353 bytes for each message.
+
+## Server implementation considerations
```

Comment:
I think server should be prepared to handle a case if an utf-8 character is split mid-way

Reply:
What do you mean "handle"? What would the expected behaviour be?

Reply:
Somehow handle. Even if by dropping the message on the floor and sending FAIL to client. Or by passing it to clients as is.

I think this shouldn't be left to implementation details to every implementation to decide for temselves.

Reply:
I expect servers don't want to inspect the content at this level, so this will probably just involve forwarding onto clients as is. We should probably tell clients not to split within multibyte characters though.

Reply:
e8efcaecb96912df1d73b2f5b9d9b51b9a09f5ed